File: CSharp/DesignPatterns/Iterator.cs
// Iterator Pattern in C#
using System;

public class DoublyLinkedList<T>
{
    public class Node
    {
        public T value;
        public Node next;
        public Node prev;

        public Node(T value, Node next, Node prev)
        {
            this.value = value;
            this.next = next;
            this.prev = prev;
        }
    }

    protected Node head;
    protected Node tail;

    public int Size { get; private set; } = 0;

    public DoublyLinkedList()
    {
        head = null;
        tail = null;
        Size = 0;
    }

    public bool IsEmpty()
    {
        return Size <= 0;
    }

    /*
        For this example we'll skip some of the nitty gritty methods
        in order to focus on just the iterator.
    */

    // Insertion method:
    public void Push(T value)
    {
        if(this.IsEmpty())
        {
            Node tmp = new Node(value, null, null);
            this.head = tmp;
            this.tail = tmp;
            this.Size++;
            return;
        }
        else
        {
            Node tmp = this.tail;
            tmp.next = new Node(value, null, tmp);
            tail = tmp.next;
            Size++;
        }
    }

    // ====================== //
    // Iterators!             //
    // ====================== //

    // Forward Iterator:

    public DLinkIterator GetForwardIterator()
    {
        return new DLinkIterator(this.head);
    }

    // The Forward Iterator class:
    public class DLinkIterator
    {
        Node current;

        // Accepts a node starting point:
        public DLinkIterator(Node start)
        {
            current = start;
        }

        public bool HasNext()
        {
            return current != null;
        }

        public T Next()
        {
            if(!HasNext())
            {
                throw new System.IndexOutOfRangeException("No Nodes to iterate.");
            }
            T val = current.value;
            current = current.next;
            return val;
        } 
    }

    // Reverse Iterator:
    public DLinkReverseIterator GetReverseIterator()
    {
        return new DLinkReverseIterator(this.tail);
    }

    public class DLinkReverseIterator
    {
        Node current;

        public DLinkReverseIterator(Node start)
        {
            this.current = start;
        }

        // This might be a little confusing at first:
        // In the context of our iterator, the "Next" node is actually the previous!
        public bool HasNext()
        {
            return current != null;
        }

        public T Next()
        {
            if(!HasNext())
            {
                throw new System.IndexOutOfRangeException("No more nodes to iterate!");
            }

            T val = current.value;
            current = current.prev;
            return val;
            
        }
    }
}


/* 
   ==========================
     Demo of the Iterators:
   ========================== 
*/
public class Solution
{
    public static void Main(string[] args)
    {
        // Construct a DLink:

        DoublyLinkedList<string> myList = new DoublyLinkedList<string>();

        myList.Push("Hello");
        myList.Push("World");
        myList.Push("Goodbye");
        myList.Push("World");

        // Iterate the list forwards:
        DoublyLinkedList<string>.DLinkIterator forwardIter = 
        myList.GetForwardIterator();

        while(forwardIter.HasNext())
        {
            Console.WriteLine(forwardIter.Next());
        }

        /*
            Output:
                Hello
                World
                Goodbye
                World
        */

        Console.WriteLine("=======");


        DoublyLinkedList<string>.DLinkReverseIterator reverseIter = 
        myList.GetReverseIterator();

        while(reverseIter.HasNext())
        {
            Console.WriteLine(reverseIter.Next());
        }

        /*
            Output:
                World
                Goodbye
                World
                Hello
        */
    }
}

File: CPlusPlus/DesignPatterns/Iterator.cpp
#include <stdexcept>
#include <iostream>

// DLink Node class:
class Node {
    public:
        Node(int value)
            : value(value), next(nullptr), prev(nullptr)
        {}
        Node(int value, Node* next, Node* prev)
            : value(value), next(next), prev(prev)
        {}
        int value;
        Node* next;
        Node* prev;
};

// =====================
// Iterators!
// =====================

// Forward Iterator:
class DLinkIterator
{
    public:
        DLinkIterator(Node* start)
        : current(start)
        {};
        bool HasNext() { return this->current != nullptr; }
        int Next();

    protected:
        Node* current;
};

int DLinkIterator::Next()
{
    if(!this->HasNext())
    {
        throw std::out_of_range("No more nodes to iterate!");
    }

    int val = this->current->value;
    this->current = this->current->next;

    return val;
}

// Reverse Iterator:
class DLinkReverseIterator
{
    public:
        DLinkReverseIterator(Node* start)
        : current(start)
        {};
        bool HasNext() { return this->current != nullptr; }
        int Next();

    protected:
        Node* current;
};

int DLinkReverseIterator::Next()
{
    if(!this->HasNext())
    {
        throw std::out_of_range("No more nodes to iterate!");
    }

    int val = this->current->value;
    // Iterate backwards!
    this->current = this->current->prev;
    return val;
}


// =============================================
// Dumbed-Down version of a Doubly-Linked-List:
// =============================================
class DoublyLinkedList {
    public:
        DoublyLinkedList()
            : size(0), head(nullptr), tail(nullptr)
        {}
        ~DoublyLinkedList();
        bool IsEmpty();
        void Add(int value);
        int Length();

        DLinkIterator GetIterator() { return DLinkIterator(this->head); };
        DLinkReverseIterator GetReverseIterator() { return DLinkReverseIterator(this->tail); };

    private:
        Node* head;
        Node* tail;
        int size;
};

// DLink Implementation:
DoublyLinkedList::~DoublyLinkedList() {
    Node* tmp = this->head;
    this->tail = nullptr;
    while(tmp != nullptr) {
        Node* deadPtr = tmp;
        tmp = tmp->next;
        delete tmp;
    }
    this->head = nullptr;
}

bool DoublyLinkedList::IsEmpty() {
    return this->size == 0;
}

void DoublyLinkedList::Add(int value) {
    if(this->IsEmpty()) {
        Node* tmp = new Node(value);
        this->head = tmp;
        this->tail = tmp;
        ++this->size;
        return;
    }
    else {
        Node* tmp = this->tail;
        tmp->next = new Node(value, nullptr, tmp);
        tail = tmp->next;
        ++this->size;
    }
}

int DoublyLinkedList::Length() {
    return this->size;
}


// Demo:

int main()
{
    
    DoublyLinkedList myList = DoublyLinkedList();

    myList.Add(1);
    myList.Add(2);
    myList.Add(3);
    myList.Add(4);
    myList.Add(5);

    // Iterate forwards:
    std::cout << "Forward iterator:" << std::endl;
    DLinkIterator it = myList.GetIterator();

    while(it.HasNext())
    {
        std::cout << it.Next() << std::endl;
    }

    /*
        Output:
            1
            2
            3
            4
            5
    */

    // Iterate backwards:
    std::cout << "Reverse iterator:" << std::endl;
    DLinkReverseIterator reverseIt = myList.GetReverseIterator();

    while(reverseIt.HasNext())
    {
        std::cout << reverseIt.Next() << std::endl;
    }

    /*
        Output:
            5
            4
            3
            2
            1
    */

}